### PR TITLE
Basic Blob

### PR DESCRIPTION
**Pull request by ThatDan123**
Adds the blob antag, is currently spawned naturally in a blob round. Admins can spawn them in non-blob rounds though.

Allows for multiple blobs but only the first one to 400 tiles will win. Currently ends the round if a blob wins, can be turned off by admin through VV

Main Issues left:
-Blob is classed as a ghost so can hear and speak on dead chat. (Not the end of the world but would be nice to fix)
-Blob spores cant move over blob structures (needs addressing somehow)
-UI could do with a touch up :p

Missing features (wont be in this PR):
-Different strains
-Rallying spores
-Blobbernaut (need proper ghost role and mind transfer system)
-Blob Zombie
-Node network (Might not be needed, but would be nice to have, means stuff needs to be near nodes to function)

CL: Added a basic blob round

### CODE REVIEW

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 1]
```diff
@@ -0,0 +1,1266 @@
...
+			}
+
+			playerSync.SetPosition(node.WorldPosServer());
+		}
+
+		#endregion
+
+		#region SyncVars
+
+		private void SyncResources(int oldVar, int newVar)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:363]: Codacy found an issue: [Remove this unused method parameter 'oldVar'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

**Comment by Bod9001** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- UnityProject/Assets/Scripts/UI/Systems/MainHUD/UI Bottom/UI_Blob.cs  2
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobStructure.cs  1
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/Blob.cs  1
- UnityProject/Assets/Scripts/Managers/GameManager.Escape.cs  3
- UnityProject/Assets/Scripts/Systems/Antagonists/Objectives/BlobDestroyStation.cs  2
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs  19
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobStarter.cs  9
- UnityProject/Assets/Scripts/Core/Input System/BlobMouseInputController.cs  8
         

Clones added
============
- UnityProject/Assets/Scripts/UI/Systems/ControlDisplays.cs  3
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobStarter.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029&bid=20637471)

File: UnityProject/Assets/Scripts/Core/Input System/BlobMouseInputController.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		}
+	}
+
+	private bool ClicksFromBlobHud()
+	{
+		if (placeOther)
+		{
+			blobPlayer.CmdToggleRemove(true);
+
+			switch (blobConstructs)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Core/Input System/BlobMouseInputController.cs:85]: Codacy found an issue: [Add a 'default' clause to this 'switch' statement.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 3]
```diff
@@ -0,0 +1,1266 @@
...
+		{
+			worldPos.z = 0;
+
+			if (!ValidateAction(worldPos)) return;
+
+			GameObject prefab = null;
+
+			var cost = 0;
+
+			switch (blobConstructs)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:768]: Codacy found an issue: [Add a 'default' clause to this 'switch' statement.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 4]
```diff
@@ -0,0 +1,1266 @@
...
+			if (BlobRemoveMode)
+			{
+				InternalRemoveBlob(worldPos);
+				return;
+			}
+
+			PlaceBlobOrAttack(worldPos);
+		}
+
+		private bool PlaceBlobOrAttack(Vector3Int worldPos, bool autoExpanding = false)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:411]: Codacy found an issue: [Change return type to 'void'; not a single caller uses the returned value.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 5]
```diff
@@ -0,0 +1,1266 @@
...
+			resourceText.text = newVar.ToString();
+		}
+
+		private void SyncHealth(float oldVar, float newVar)
+		{
+			health = newVar;
+			healthText.text = newVar.ToString();
+		}
+
+		private void SyncNumOfBlobTiles(int oldVar, int newVar)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:375]: Codacy found an issue: [Remove this unused method parameter 'oldVar'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Managers/CentComm.cs [hunk 6]
```diff
@@ -83,6 +83,10 @@ public void UpdateStatusDisplay(StatusDisplayChannel channel, string text)
 	private string AntagInitialUpdate =
 		"Enemy communication intercepted. Security level elevated.";
 
+	public static string BioHazardReportTemplate =
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Managers/CentComm.cs:86]: Codacy found an issue: [Change the visibility of 'BioHazardReportTemplate' or make it 'const' or 'readonly'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 7]
```diff
@@ -0,0 +1,1266 @@
...
+			{
+				if (blob != null && blob.isNormal)
+				{
+					if (blobConstructs != BlobConstructs.Node && !ValidateDistance(worldPos, true))
+					{
+						Chat.AddExamineMsgFromServer(gameObject, $"Too close to another factory or resource blob, place at least {buildDistanceLimit}m away");
+						return;
+					}
+
+					if (blobConstructs == BlobConstructs.Node && !ValidateDistance(worldPos, false, true, true))
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:796]: Codacy found an issue: [Remove this default value assigned to parameter 'excludeNodes'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 8]
```diff
@@ -0,0 +1,1266 @@
...
+
+		#region SyncVars
+
+		private void SyncResources(int oldVar, int newVar)
+		{
+			resources = newVar;
+			resourceText.text = newVar.ToString();
+		}
+
+		private void SyncHealth(float oldVar, float newVar)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:369]: Codacy found an issue: [Remove this unused method parameter 'oldVar'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs (hunk 1 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:363]: Needed

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs (hunk 4 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:411]: It is used...

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs (hunk 5 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:375]: Needed

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs (hunk 8 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:369]: Needed

**Comment by Bod9001** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- UnityProject/Assets/Scripts/Systems/Antagonists/Objectives/BlobDestroyStation.cs  2
- UnityProject/Assets/Scripts/UI/Systems/MainHUD/UI Bottom/UI_Blob.cs  2
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs  19
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobStructure.cs  1
- UnityProject/Assets/Scripts/Core/Input System/BlobMouseInputController.cs  8
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobStarter.cs  9
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/Blob.cs  1
- UnityProject/Assets/Scripts/Managers/GameManager.Escape.cs  3
         

Clones added
============
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobStarter.cs  1
- UnityProject/Assets/Scripts/UI/Systems/ControlDisplays.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029&bid=20637471)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 9]
```diff
@@ -0,0 +1,1272 @@
...
+
+					var result = Spawn.ServerPrefab(prefab, worldPos, gameObject.transform);
+
+					if (!result.Successful) return;
+
+					Chat.AddExamineMsgFromServer(gameObject, $"You grow a {blobConstructs} blob.");
+
+					var structure = result.GameObject.GetComponent<BlobStructure>();
+
+					switch (blobConstructs)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:818]: Codacy found an issue: [Add a 'default' clause to this 'switch' statement.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 10]
```diff
@@ -0,0 +1,1272 @@
...
+			}
+
+			playerSync.SetPosition(node.WorldPosServer());
+		}
+
+		#endregion
+
+		#region SyncVars
+
+		private void SyncResources(int oldVar, int newVar)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:366]: Codacy found an issue: [Remove this unused method parameter 'oldVar'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 11]
```diff
@@ -0,0 +1,1272 @@
...
+			if (BlobRemoveMode)
+			{
+				InternalRemoveBlob(worldPos);
+				return;
+			}
+
+			PlaceBlobOrAttack(worldPos);
+		}
+
+		private bool PlaceBlobOrAttack(Vector3Int worldPos, bool autoExpanding = false)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:414]: Codacy found an issue: [Change return type to 'void'; not a single caller uses the returned value.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 12]
```diff
@@ -0,0 +1,1272 @@
...
+			resourceText.text = newVar.ToString();
+		}
+
+		private void SyncHealth(float oldVar, float newVar)
+		{
+			health = newVar;
+			healthText.text = newVar.ToString();
+		}
+
+		private void SyncNumOfBlobTiles(int oldVar, int newVar)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:378]: Codacy found an issue: [Remove this unused method parameter 'oldVar'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 13]
```diff
@@ -0,0 +1,1272 @@
...
+
+		#region SyncVars
+
+		private void SyncResources(int oldVar, int newVar)
+		{
+			resources = newVar;
+			resourceText.text = newVar.ToString();
+		}
+
+		private void SyncHealth(float oldVar, float newVar)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:372]: Codacy found an issue: [Remove this unused method parameter 'oldVar'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

**Comment by ThatDan123** [comment]
I probably wont be fixing the issues in this PR as the rest of the mechanics could use a staging test, and if its broken then can just remove it from the round pool. But currently from my testing it should be fully playable.

The issues will probably take sometime to fix and if anyone wants to help out please say :p As this really isnt my expertise changing core systems (Dealing with movement and if we attempt to move over to being a player prefab instead of ghost)

**Comment by Bod9001** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- UnityProject/Assets/Scripts/Core/Input System/BlobMouseInputController.cs  8
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobStarter.cs  9
- UnityProject/Assets/Scripts/Managers/GameManager.Escape.cs  3
- UnityProject/Assets/Scripts/UI/Systems/MainHUD/UI Bottom/UI_Blob.cs  2
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/Blob.cs  1
- UnityProject/Assets/Scripts/Systems/Antagonists/Objectives/BlobDestroyStation.cs  2
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs  19
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobStructure.cs  1
         

Clones added
============
- UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobStarter.cs  1
- UnityProject/Assets/Scripts/UI/Systems/ControlDisplays.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029&bid=20637471)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 14]
```diff
@@ -0,0 +1,1275 @@
...
+			}
+
+			playerSync.SetPosition(node.WorldPosServer());
+		}
+
+		#endregion
+
+		#region SyncVars
+
+		private void SyncResources(int oldVar, int newVar)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:366]: Codacy found an issue: [Remove this unused method parameter 'oldVar'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 15]
```diff
@@ -0,0 +1,1275 @@
...
+			if (BlobRemoveMode)
+			{
+				InternalRemoveBlob(worldPos);
+				return;
+			}
+
+			PlaceBlobOrAttack(worldPos);
+		}
+
+		private bool PlaceBlobOrAttack(Vector3Int worldPos, bool autoExpanding = false)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:414]: Codacy found an issue: [Change return type to 'void'; not a single caller uses the returned value.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 16]
```diff
@@ -0,0 +1,1275 @@
...
+			resourceText.text = newVar.ToString();
+		}
+
+		private void SyncHealth(float oldVar, float newVar)
+		{
+			health = newVar;
+			healthText.text = newVar.ToString();
+		}
+
+		private void SyncNumOfBlobTiles(int oldVar, int newVar)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:378]: Codacy found an issue: [Remove this unused method parameter 'oldVar'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

File: UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs [hunk 17]
```diff
@@ -0,0 +1,1275 @@
...
+
+		#region SyncVars
+
+		private void SyncResources(int oldVar, int newVar)
+		{
+			resources = newVar;
+			resourceText.text = newVar.ToString();
+		}
+
+		private void SyncHealth(float oldVar, float newVar)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Antagonists/Antags/Blob/BlobPlayer.cs:372]: Codacy found an issue: [Remove this unused method parameter 'oldVar'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6357029)

**Pull request merged**